solutions/gold/cses-2416.mdx

```diff
@@ -2,106 +2,185 @@
 id: cses-2416
 source: CSES
 title: Increasing Array Queries
-author: Andi Qu
+author: Andi Qu, Kevin Sheng
 ---
 
-**Time Complexity:** $\mathcal O((N + Q) \log N)$.
+<Spoiler title="Hint 1">
+
+Can you first come up with an algorithm for a single query?
+It doesn't have to be crazy efficient; $\mathcal{O}(N)$ time is fine.
+
+</Spoiler>
+
+<Spoiler title="Answer to Hint 1">
+
+Let's iterate through the array in order and keep track of the largest element $m$ we've seen so far.
+For each element $x$, we first update $m$.
+Then, if the current one is smaller than $m$, we add $m-x$ to the total
+
+</Spoiler>
+
+<Spoiler title="Hint 2">
+
+Since we can't process each query like this, let's try to come up with the concept
+of a "contribution" of a certain element.
+
+Take the array $[10,4,11,3]$, for example.
+Try to think of how $10$ and $11$ in particular contribute to the answer.
 
-In this problem, we can answer the queries offline (i.e. read in the queries and
-process them in a different order). More specifically, we process queries in
-order of their left endpoints.
+</Spoiler>
 
-First, let's think about how we'd answer a single query $(l, r)$. Consider the
-following algorithm:
+<Spoiler title="Solution">
 
-- Let $\texttt{mx}_0 = 0$ and $j = 0$.
-- For each $i$ in $[l, r]$:
-  - If $x_i > \texttt{mx}_j$, then set $\texttt{mx}_{j + 1} = x_i$ and increment
-    $j$.
-  - Otherwise, add $\texttt{mx}_j - x_i$ to the answer.
+## Explanation
 
-Notice how each $\texttt{mx}_j$ (except the last) contributes a fixed amount to
-the answer, regardless of $r$. If $\texttt{pref}$ is the prefix sum array of $x$
-and $\texttt{idx}_j$ is the index of $\texttt{mx}_j$ in $x$, then this amount is
+### Setup
+
+Let's follow the line of thought from Hint 2.
+Say we have an element of value $x$ at index $i$, and the next element strictly greater than $x$ is at index $j$.
+If there is no next greater element, let $j$ be one greater than the index of the final element.
+
+The contribution from $x$ to a query would be as follows:
 
 $$
-(\texttt{idx}_{j + 1} - 1 - \texttt{idx}_j) \cdot \texttt{mx}_j - (\texttt{pref}_{\texttt{idx}_{j + 1} - 1} - \texttt{pref}_{\texttt{idx}_j})
+x \cdot (j-i) - \sum_{k=i}^{j-1} \texttt{arr}[k]
 $$
 
-This means that for a fixed $l$, we can compute each $\texttt{mx}_j$ and their
-contributions, and then use `std::upper_bound` and a BIT to answer queries with
-variable $r$!
+The second term is the sum of all elements in the interval $[i,j-1)$.
```

Comment:
Should this be [i, j)? (the summation goes from i to j - 1 right)

Reply:
oh yeah mbmb